### PR TITLE
Catch ValueError when calling memoryview()

### LINKED ISSUE #5401: [BUG] Fused memoryviews returns `ValueError` instead of `TypeError`.
### Describe the bug

Fused memoryviews returns `ValueError` instead of `TypeError`.

### Code to reproduce the behaviour:

```cython
cimport numpy as cnp
import cython

ctypedef fused np_numeric_t:
    cnp.int8_t
    cnp.int16_t
    cnp.int32_t
    cnp.int64_t
    cnp.uint8_t
    cnp.uint16_t
    cnp.uint32_t
    cnp.uint64_t
    cnp.float32_t
    cnp.float64_t
    cnp.longdouble_t
    cnp.complex64_t
    cnp.complex128_t

@cython.initializedcheck(False)
def bandwidth_c(np_numeric_t[:, ::1]A):
    return
```

## Cython 3
```python
>>> import numpy as np
>>> zz = np.zeros([5, 5], dtype='M')
>>> import fused
>>> fused.bandwidth_c(zz)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "fused.pyx", line 31, in fused.__pyx_fused_cpdef
    @cython.initializedcheck(False)
ValueError: cannot include dtype 'M' in a buffer
```
## Cython 0.29.X:

```python
>>> import numpy as np
>>> zz = np.zeros([5, 5], dtype='M')
>>> import fused
>>> fused.bandwidth_c(zz)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "fused.pyx", line 32, in fused.__pyx_fused_cpdef
    def bandwidth_c(np_numeric_t[:, ::1]A):
TypeError: No matching signature found
```

### OS

MacOS

### Python version

3.9.16

### Cython version

master

### Additional context

See discussion in mailing list: https://mail.python.org/pipermail/cython-devel/2023-April/005423.html

### PR DESCRIPTION
**Pull request by matusvalo**
Fixes #5401

### CODE REVIEW

**Comment by matusvalo** [comment]
Implementation based on comment https://github.com/cython/cython/issues/5401#issuecomment-1520501523

File: tests/memoryview/numpy_memoryview.pyx [hunk 1]
```diff
@@ -734,3 +734,18 @@ def same_type_after_arrays_composite():
     cdef SameTypeAfterArraysStructComposite element
     arr = np.ones(2, np.asarray(<SameTypeAfterArraysStructComposite[:1]>&element).dtype)
     cdef SameTypeAfterArraysStructComposite[:] memview = arr
+
+
+ctypedef fused np_numeric_t:
```
Comment (da-woods) [inline tests/memoryview/numpy_memoryview.pyx:739]: Could we also try a fused type of `np.float64_t[:]` and `object`. In that case the `dtype="M"` array should match `object` (so not produce any error).

I'm fairly sure it'll work, but it'd be nice to have the explicit test

File: tests/memoryview/numpy_memoryview.pyx [hunk 2]
```diff
@@ -734,3 +734,18 @@ def same_type_after_arrays_composite():
...
+    np.float64_t
+
+def test_issue5401(np_numeric_t[:] A):
```
Comment (da-woods) [inline tests/memoryview/numpy_memoryview.pyx:742]: I suspect @scoder is likely to complain about the non-descriptive naming here.

I'd suggest `test_invalid_buffer_fused_memoryview` maybe?

File: tests/memoryview/numpy_memoryview.pyx (hunk 1 above)
Comment (matusvalo) [reply tests/memoryview/numpy_memoryview.pyx:738]: done thx!

File: tests/memoryview/numpy_memoryview.pyx (hunk 2 above)
Comment (matusvalo) [reply tests/memoryview/numpy_memoryview.pyx]: done

**Comment by da-woods** [comment]
Looks good to me - thanks. I'll merge it unless anyone has any comments in the next few days

**Pull request merged**